---
---

The philosophy of 'BCH709 Introduction to Bioinformatics' is to teach researchers basic concepts, skills, and tools for working with data so that student can get their own bioinformatics work done without pain. This illustrates the concept of bioinformatics analysis and Next Generation Sequencing for genomics including hands-on bioinformatics, understanding genomics through bioinformatics, use of command-line in multiple platforms, use of command-line tools to analyze sequencing data, and connecting to and using cloud computing.


> ## Class Schedule
> MonWed 9:00AM - 10:15AM  
> Aug 24, 2020 - Dec 18 2020
{: .prereq}


> ## Modes of Instruction
> Response to COVID-19, BCH709 will offer Alternative HyFlex course model.
>
> **Alternative HyFlex**: Course originally scheduled to be delivered in-person, but will have attendance adjustments so students will alternate in-person and remote participation.  Students will receive communication directly from their instructors with specifics. Fully remotely participation by students can be accommodated for those with extenuating circumstances. 
>
> **In addition, we will provide recorded video for everyone.**
{: .callout}


> ## Class Room location
> Classroom was changed to Fleischmann Agriculture Building room 234 (FA234)
> ![Classroom location](./fig/classroom_location2.png){: width="50%" height="50%"}
> 
> [FA234](https://www.google.com/maps/place/39%C2%B032'15.7%22N+119%C2%B048'43.5%22W/@39.5385648,-119.8114072,17.67z/data=!4m5!3m4!1s0x0:0x0!8m2!3d39.5376946!4d-119.81208)  
{: .prereq}

> ## Class Room location
> Alternative participate via Zoom
> Join Zoom Meeting
> [Zoom link](https://unr.zoom.us/j/6314611123?pwd=SE5ZOGNlcjIrZEszcnl6cElTQ3U1Zz09)
> Meeting ID: 631 461 1123
> Passcode: 709
> One tap mobile
> +16699006833,,6314611123# US (San Jose)
> +12532158782,,6314611123# US (Tacoma)
>
{: .prereq}


> ## Prerequisites
> - Computer with ethernet port or wifi (If in case you bring your **desktop**, please do not bring your monitor. we have a monitor in our classroom)
> Online introduction to Linux. Students must complete one of the following online tutorials (or both) before class begins.
> - UNR affilated email **\<ID\>@unr.edu or \<ID\>@nevada.unr.edu** - [How to Activate](https://oit.unr.edu/services-and-support/login-ids-and-passwords/netid/netid-activation/)
> - Please fill this [form](https://forms.gle/91784nfQnj2QweHS9)
> - [Please apply your Pronghorn account here](https://nevada.formstack.com/forms/high_performance_computing_application) ; [Please follow this instrunction](https://plantgenomicslab.github.io/BCH709/pronghorn/index.html)
> - [Setup your computer](https://plantgenomicslab.github.io/BCH709/setup.html)
> - [Setup Slack ID](https://unrrc.slack.com/)
> - Please register [datacamp](https://www.datacamp.com/)
{: .callout}


> ## Course description
>As contemporary biologists we have entered an age where the use of computers in our daily work has become all but essential. The manipulation and analysis of DNA, RNA, and protein data by electronic means has become a routine task. Further, the amount of DNA, RNA and protein sequence data we are putting into databases every day is expanding at a geometric rate, and with coming advances in sequencing technology, this rate is only expected to increase. With all this new data, analysis by individual humans is simply not possible. Thus, in the past 15 years, computational biology has emerged as a field concerned with storage, manipulation, and extraction of valuable information from all this new data. However, because computational biology is an emerging field, organized courses are generally saved for higher-level study, and often are not required parts of an undergraduate curriculum. We seek to fill this void in education and create a course that will introduce students to bioinformatics at an earlier point in their education. This knowledge will prove to be not simply useful, but essential, for any student considering a degree in any area of biology and medical science.
{: .prereq}



## Tentative Course Schedules

| Week             | Date | Topics        |
|--------------|------|-------------------------------------------------|
| Week1        | 8/24 | [Orientation](https://plantgenomicslab.github.io/BCH709/Orientation/index.html)|
| Week1        | 8/26 | [Intro Bioinformatics](https://plantgenomicslab.github.io/BCH709/Intro/index.html) |
| Week2        | 8/31 |  [Linux Enviroment](https://plantgenomicslab.github.io/BCH709/Linux_Enviroment_and_command_line/index.html) / [Linux command line](https://plantgenomicslab.github.io/BCH709/Linux_Enviroment_and_command_line2/index.html)|
| Week2        | 9/2  | [Linux command line](https://plantgenomicslab.github.io/BCH709/Linux_Enviroment_and_command_line2/index.html)|
| Week3        | 9/7  | LABOR DAY|
| Week3        | 9/9  | [Linux Enviroment and command line](https://plantgenomicslab.github.io/BCH709/Linux_Enviroment_and_command_line3/index.html)|
| Week4        | 9/14 |[Linux Enviroment and command line](https://plantgenomicslab.github.io/BCH709/Linux_Enviroment_and_command_line3/index.html) / [Github / Compile / Conda](https://plantgenomicslab.github.io/BCH709/Github_compile_conda/index.html)  |
| Week4        | 9/16 | [Github / Compile / Conda](https://plantgenomicslab.github.io/BCH709/Github_compile_conda/index.html)  |
| Week5        | 9/21 |  [Sequencing methods and strategies](https://www.dropbox.com/s/5ajz4e6kd9twxux/Sequencing.pdf) / [RNA-Seq](https://plantgenomicslab.github.io/BCH709/RNASeq/index.html)|
| Week5        | 9/23 |  [RNA-Seq](https://plantgenomicslab.github.io/BCH709/RNASeq/index.html) / [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly/index.html)|
| Week6        | 9/28 |  [RNA-Seq](https://plantgenomicslab.github.io/BCH709/RNASeq/index.html) / [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly/index.html) / [DEG analysis](https://plantgenomicslab.github.io/BCH709/DEG2/index.html) |
| Week6        | 9/30  | [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly/index.html)|
| Week7        | 10/5  | MIDTERM EXAM
| Week7        | 10/7 | [Introduction of R R plotting (Dr. Tong Zhou)](https://plantgenomicslab.github.io/BCH709/Intro_R/index.html)|
| Week8        | 10/12 | [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly3/index.html) |
| Week8        | 10/14 | [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly3/index.html) |
| Week9        | 10/19 | [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly4/index.html)|
| Week9       | 10/21 | [Transcriptome assembly](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly5/index.html) [DEG](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly4/index.html) |
| Week10       | 10/26  |[DEG](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly6/index.html) |
| Week10       | 10/28 | [DEG](https://plantgenomicslab.github.io/BCH709/transcriptome_assembly7/index.html) [Genome assembly](https://plantgenomicslab.github.io/BCH709/Genome_assembly_1/index.html) |
| Week11       | 11/2  | [Genome assembly](https://plantgenomicslab.github.io/BCH709/Genome_assembly_1/index.html)  | [Genome structure](https://plantgenomicslab.github.io/BCH709/Genome_annotation/index.html)|
| Week11       | 11/4 | [Genome assembly](https://plantgenomicslab.github.io/BCH709/Genome_assembly_2/index.html) |
| Week12       | 11/9 |Gene family analysis and phylogenetics  (Dr. David Alvarez-Ponce)   |
| Week12       | 11/11 | Vertans Day |
| Week13       | 11/16 | [Genome assembly](https://plantgenomicslab.github.io/BCH709/Genome_assembly_2_1/index.html)
| Week13       | 11/18 | [Genome annotation and annotation](https://plantgenomicslab.github.io/BCH709/Genome_annotation/index.html)|
| Week14       | 11/23 | [RNA-Seq tutorial](https://plantgenomicslab.github.io/BCH709/Genome_annotation/RNASeq_tutorials.html)                          |
| Week14       | 11/25 | [Enrichment analysis](https://plantgenomicslab.github.io/BCH709/Genome_description/index.html)|
| Week15       | 11/30  |  Transcriptome analysis |
| Week15 | 12/2 |  Transcriptome analysis   |



> ## Getting Started
>
> This lesson assumes that learners have no prior experience with the tools covered in the workshop. 
> However, learners are expected to have some familiarity with biological concepts,
> including the concept of genomic variation within a population. Participants should bring their own laptops and plan to participate actively. 
> 
> To get started, follow the directions in the [Setup](setup.html) tab to 
> get access to the required software and data for this workshop.
> 
{: .prereq}
<!-- 
> ## Data
> 
> This workshop uses data from a long term evolution experiment published in 2016: [Tempo and mode of genome evolution in a 50,000-generation experiment](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/) by Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL, Dasgupta A, Wu GC, Wielgoss S, Cruveiller S, Médigue C, Schneider D, and Lenski RE. (doi: 10.1038/nature18959)
>
> All of the data used in this workshop can be [downloaded from Figshare](https://figshare.com/articles/Data_Carpentry_Genomics_beta_2_0/7726454). 
> More information about this data is available on the [Data page](https://datacarpentry.org/organization-genomics/data/).
{: .prereq} 
-->



## Optional Additional Meeting

### _Research Computing Hackathon_  ([Hosted by HPC team](https://www.unr.edu/research-computing/hpc))  
Every Friday at 2:00pm to 4:00pm through SLACK

[Hackathons](https://en.wikipedia.org/wiki/Hackathon) provide a space for hands-on training and solution development within a Research Computing environment at the University. This is also a place to get clarification on questions/concerns regarding the HPC environment. Please bring problems to challenge the HPC team, the Office of Information Technology, and research colleagues. If you don't need help, we still encourage you to attend and share your time and expertise with those in need of assistance. You don’t need to be an expert to attend a hackathon. Individuals at all computing skill levels are welcome! Won Yim will attend this hackathon.

### _Meeting_  
Office [Howard Medical Science 216](https://goo.gl/maps/o41BMmcawsTPoES57)  
I prefer to have online meeting through SLACK.
  
## Optional Reading materials
- [Introduction to Bioinformatics (3rd Edition)](http://app.knovel.com/web/toc.v/cid:kpIBE00007/viewerType:toc/ "Introduction to Bioinformatics (3rd Edition)")

- [Learn Linux Shell Scripting - Fundamentals of Bash 4.4](https://learning.oreilly.com/library/view/learn-linux-shell/9781788995597/ "Learn Linux Shell Scripting - Fundamentals of Bash 4.4")

- [Effective awk Programming, 3rd Edition](https://learning.oreilly.com/library/view/effective-awk-programming/0596000707/ "Effective awk Programming, 3rd Edition")

- [Bioinformatics with Python Cookbook - Second Edition](https://learning.oreilly.com/library/view/bioinformatics-with-python/9781789344691/ "Bioinformatics with Python Cookbook - Second Edition")

- [Basic Applied Bioinformatics](https://learning.oreilly.com/library/view/basic-applied-bioinformatics/9781119244332/ "Basic Applied Bioinformatics")

- [Ubuntu Unleashed 2019 Edition: Covering 18.04, 18.10, 19.04, 13/e](https://learning.oreilly.com/library/view/ubuntu-unleashed-2019/9780134985497/ "Ubuntu Unleashed 2019 Edition: Covering 18.04, 18.10, 19.04, 13/e")

- [Unix and Perl: Keith Bradnam & Ian Korf](https://j.p.gogarten.uconn.edu/mcb5472_2018/current.pdf)

- [Other bioinformatics books from knovel](http://app.knovel.com/web/search.v?q=bioinformatics&search_type=tech-reference&rows=10&offset=0&group_by=true&my_subscription=true&sort_on=default&content_type=all_references&include_synonyms=no "Other bioinformatics books from knovel")

- [Other bioinformatics books from Oreilly](https://learning.oreilly.com/search/?query=bioinformatics&extended_publisher_data=true&highlight=true&include_assessments=false&include_case_studies=true&include_courses=true&include_orioles=true&include_playlists=true&include_collections=false&include_notebooks=false&is_academic_institution_account=false&sort=relevance&facet_json=true "Other bioinformatics books from Oreilly")

**Note: all reading material can be freely accessed and downloaded from the UNR internet.**

## Other Website
[Command line website](http://35.199.189.11:8081/  "Command line website")  
[Command-line bootcamp-server2](http://34.82.237.168:8081/)

[Command-line bootcamp-server3](http://35.247.103.91:8081/)

[Command-line bootcamp-server4](http://34.83.44.229:8081/)

[Plant Genomics Lab](https://www.plantbioinformatics.org/ "Plant Genomics Lab")

  
> ## Frequently Asked Questions  
Read our [FAQ](./_episodes/FAQ/FAQ.md). Currently, this page is empty, but we will build it through the class.
{: .prereq}

  
>## Teaching Platform
This lecture was designed to be run on [Unix-base system](https://en.wikipedia.org/wiki/Unix) such as 
Ubuntu, mac, etc. All the software and data used in the class will be open source. All example data will be hosted on a Google Cloud Service. If you want to know how to use Unix-base system on your computer, please follow the directions in the [Setup](setup.html) tab.
{: .prereq}

  

The website theme was adapted from the original by [Data Carpentry](https://datacarpentry.org/). The infrastructure, including adventure-time and docker-browser-server, was built by @maxogden and @mafintosh. The setup of this app was based on the get-data adventure. This adventure app was made by Richard Smith-Unna. The lecture materials were crafted by Won Yim. This work is licensed under a Creative Commons 4.0 International License.
