# Human DEG
## Activate environment
```bash
conda activate BCH709_RNASeq
```
## Copy Mouse featurecount results
```bash
cp /data/gpfs/assoc/bch709-3/Course_materials/human/human_featurecount.txt  /data/gpfs/assoc/bch709-3/${USER}/human/readcount/

```

## Clean sample name
```bash
cut -f1,7-  human_featurecount.txt |  egrep -v "#" | sed 's/\Aligned\.sortedByCoord\.out\.bam//g; s/\.bam//g' >> human_featurecount_only.cnt
cut -f1,6-  human_featurecount.txt |  egrep -v "#" | sed 's/\Aligned\.sortedByCoord\.out\.bam//g; s/\.bam//g' >> human_featurecount_length.cnt
```

## Copy read count to DEG folder
 ```bash
cp /data/gpfs/assoc/bch709-3/${USER}/human/readcount/human_featurecount_* /data/gpfs/assoc/bch709-3/${USER}/human/DEG
```


## Go to DEG folder
```bash
cd /data/gpfs/assoc/bch709-3/${USER}/human/DEG
```


### sample files
```bash
cp /data/gpfs/assoc/bch709-3/Course_materials/human/human_sample_association /data/gpfs/assoc/bch709-3/${USER}/human/DEG

cp /data/gpfs/assoc/bch709-3/Course_materials/human/human_contrast /data/gpfs/assoc/bch709-3/${USER}/human/DEG

cd /data/gpfs/assoc/bch709-3/${USER}/human/DEG
```

## PtR (Quality Check Your Samples and Biological Replicates)

Once you've performed transcript quantification for each of your biological replicates, it's good to examine the data to ensure that your biological replicates are well correlated, and also to investigate relationships among your samples. If there are any obvious discrepancies among your sample and replicate relationships such as due to accidental mis-labeling of sample replicates, or strong outliers or batch effects, you'll want to identify them before proceeding to subsequent data analyses (such as differential expression). 
```bash
cd /data/gpfs/assoc/bch709-3/${USER}/human/DEG
PtR  --matrix human_featurecount_only.cnt  --samples human_sample_association --CPM  --log2 --min_rowSums 10   --sample_cor_matrix --compare_replicates

```

## PtR download on local
```bash
scp [YOURID]@pronghorn.rc.unr.edu:/data/gpfs/assoc/bch709-3/[YOURID]/human/DEG/*.pdf .
```

## If this is not working on local Mac
```bash
echo "setopt nonomatch" >> ~/.zshrc
```

## DEG analysis on Pronghorn
```bash
cd /data/gpfs/assoc/bch709-3/${USER}/human/DEG

run_DE_analysis.pl --matrix human_featurecount_only.cnt --method DESeq2 --samples_file human_sample_association --contrasts human_contrast --output human_rnaseq
```

## DEG analysis results
```bash
ls human_rnaseq
```

## TPM/FPKM calculation
```bash
python /data/gpfs/assoc/bch709-3/Course_materials/script/tpm_raw_exp_calculator.py -count human_featurecount_length.cnt
```

### TPM and FPKM calculation output
```bash
human_featurecount_length.cnt.fpkm.xls
human_featurecount_length.cnt.fpkm.tab
human_featurecount_length.cnt.tpm.xls
human_featurecount_length.cnt.tpm.tab
```

### DEG subset
```bash
cd /data/gpfs/assoc/bch709-3/${USER}/human/DEG/human_rnaseq
analyze_diff_expr.pl --samples /data/gpfs/assoc/bch709-3/${USER}/human/DEG/human_sample_association  --matrix /data/gpfs/assoc/bch709-3/${USER}/human/DEG/human_featurecount_length.cnt.tpm.tab -P 0.01 -C 2 --output human_RNASEQ_P001_C2
analyze_diff_expr.pl --samples /data/gpfs/assoc/bch709-3/${USER}/human/DEG/human_sample_association  --matrix /data/gpfs/assoc/bch709-3/${USER}/human/DEG/human_featurecount_length.cnt.tpm.tab -P 0.01 -C 1 --output human_RNASEQ_P001_C1
```


## DEG download
```bash
scp -r [YOURID]@pronghorn.rc.unr.edu:/data/gpfs/assoc/bch709-3/wyim/human/DEG/ .